src/crush/mapper.c

```diff
@@ -823,19 +992,46 @@ int crush_do_rule(const struct crush_map *map,
 	int vary_r = map->chooseleaf_vary_r;
 	int stable = map->chooseleaf_stable;
 
+	struct crush_work* cw;
+	int *a;
+	int *b;
+	int *c;
+
+
 	if ((__u32)ruleno >= map->max_rules) {
 		dprintk(" bad ruleno %d\n", ruleno);
 		return 0;
 	}
 
 	rule = map->rules[ruleno];
 	result_len = 0;
+
+
+	/* I'm not sure if this is necessary, or if we should always
+	   require the caller to provide workspace and error/crash if
+	   they don't. It seemed harmless to include it and
+	   potentially useful, but we could get rid of it. */
+
+	if (cwin)
+		cw = cwin;
+	else
+		cw = kmalloc(map->working_size, GFP_KERNEL);
+
+	if (!cw)
+		return 0;
+
```

Comment:
No need for kmalloc or error/crash part - we should always require the caller to provide workspace and just assume it's there.